substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1920,6 +1922,47 @@ pub mod pallet {
 
 			Ok(Pays::No.into())
 		}
+
+		/// Updates a batch of controller accounts to their corresponding stash account if they are
+		/// not the same. Ignores any controller accounts that do not exist, and does not operate if
+		/// the stash and controller are already the same.
+		///
+		/// Effects will be felt instantly (as soon as this function is completed successfully).
+		///
+		/// The dispatch origin must be Root.
+		#[pallet::call_index(28)]
+		#[pallet::weight(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32))]
+		pub fn deprecate_controller_batch(
+			origin: OriginFor<T>,
+			controllers: BoundedVec<T::AccountId, T::MaxControllersInDeprecationBatch>,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			// Ignore controllers that do not exist or are already the same as stash.
+			let filtered_batch_with_ledger: Vec<_> = controllers
+				.iter()
+				.filter_map(|controller| {
+					let ledger = Self::ledger(StakingAccount::Controller(controller.clone()));
+					ledger.ok().map_or(None, |ledger| {
+						if ledger.stash != *controller {
+							Some((controller.clone(), ledger))
+						} else {
+							None
+						}
+					})
+				})
+				.collect();
+
+			// Update unique pairs.
+			for (controller, ledger) in filtered_batch_with_ledger {
+				let stash = ledger.stash.clone();
+
+				<Bonded<T>>::insert(&stash, &stash);
```

Comment:
I think that with the current defensive proof that we're using in `make_payout`, stakers with `RewardDestination::Controller` will still get paid even before the lazy migration completes.

If this batch is executed, `RewardDestination::Controller` will point to the updated  stash values.

```
RewardDestination::Controller => Self::bonded(stash)
    .map(|controller| {
        defensive!("Paying out controller as reward destination which is deprecated and should be migrated.");
        // This should never happen once payees with a `Controller` variant have been migrated.
        // But if it does, just pay the controller account.
        T::Currency::deposit_creating(&controller, amount)
```

I think we are good here, but when it comes to removing `Bonded`, we'll definitely want to be sure the lazy migration has completed.

Reply:
Rewards would not be missed though it would be paid in the wrong account, but more importantly, we will lose the controller account on-chain before we have migrated it. While I am pretty sure we will take the steps correctly, probably just good to skip deprecating controllers in cases where `RewardDestination` is not migrated yet (just for our sanity since who knows some other non relaychain runtime might be using pallet-staking and have to go through similar migration and do not migrate `RewardDestination` before deprecating controllers.).